### ISSUE TITLE
Mockito 4

### ISSUE THREAD

**Issue by mockitoguy** [mockito/mockito#1517]
Suggested scope of Mockito 4

- [x] ~Move Java version compatibility to 8+~ - implemented in Mockito 3
- [ ] Enable strict stubbing by default in JUnit4 test runners (#769)
- [x] Remove all deprecated Mockito 2 classes (https://github.com/mockito/mockito/search?q=deprecated&unscoped_q=deprecated)
- [ ] Un-incubate (or promote) incubating features
- [x] Write migration guide for existing users
  - [ ] Write ErrorProne refactoring check that automatically moves all runners to `Silent`, so that companies can gradually upgrade.
- [ ] Move Junit5 extension from subproject to core project.

Any comments or feedback?

**Comment by TimvdLippe** [comment]
First point: yes please!

Second point: I am in favor, but we have to have a good migration strategy. E.g. how can users upgrade from 2 to 3 in a gradual manner? So while I do think we should do this, I think some documentation and upgrade guides need to be written before we making any release.

**Comment by TimvdLippe** [comment]
Maybe we should also ship fresh `module-info.java` to lock down any usages of `org.mockito.internal` for Java 9 users.

**Comment by TimvdLippe** [comment]
@ChristianSchwarz 

> * removal of duplicated API's
> * introduction of new/replacement API's for stubbing and verification using Lambdas

We have discussed this and decided that these features are not ready and should not block a Mockito 3 release. I hope to get back to that soon as well.

> * allow mixing matchers and args

This is not possible without "hacks/workarounds". A fundamental change in our API is necessary to fix this. See #1264 for a working version of that.

> * allow Collections and AtomicReference as a capturing targets, remove/deprecate ArgumentCaptor

I am confused, did I miss the background for this? I am not aware of any discussion regarding these changes 😮

**Comment by TimvdLippe** [comment]
We should also be consistent with our deprecation policy and remove all currently deprecated targets in the next major breaking release: https://github.com/mockito/mockito/search?q=deprecated&unscoped_q=deprecated

**Comment by itwarilal** [comment]
Would be great if static and final method mocking is provided out of the box. It’s easy enough to do this with powerMock but why add another dep just for that.

**Comment by qualidafial** [comment]
Please consider JUnit 5 Jupiter extensions alongside JUnit 4 test rules

**Comment by szpak** [comment]
@qualidafial How could you miss even my blog post about that ;)
https://solidsoft.wordpress.com/2018/03/27/convenient-mocking-in-mockito-with-junit-5-the-official-way/

**Comment by ChristianSchwarz** [comment]
@TimvdLippe
I think @qualidafial requests to include the JUnit5 implementation in the main mockito jar. Currently we have to add it as extra dependency.

**Comment by ChristianSchwarz** [comment]
@mockitoguy 
@TimvdLippe 
> Strict stubbing turned on by default for JUnit rules and runners 

Assume the following scenario:
 - a project have 20k test using mockito 2 
 - the project wants to upgrade to mockito 3
 - after the upgrade 10k tests fail due to the new strict stubs default behaviour

How does mockito 3 ensures that 10k dont have to be adopted to the new behaviour? Will mockito3 use a different base package like "org.mockito3" to avoid failing mockito 2 tests, so you can use both versions at the same time?

**Comment by TimvdLippe** [comment]
@ChristianSchwarz Right, I have added that to the TODO-list.

I am also planning on writing an `ErrorProne` refactoring tool that automatically moves all runners to `Silent`, so that companies can gradually upgrade their codebase.

**Comment by qualidafial** [comment]
Adding on to @ChristianSchwarz's last comment: it's easier to do a major version upgrade if the package names are different, and the Maven artifact coordinates are also distinct. So you can depends on both 2.x and 3.x during transition.

**Comment by TimvdLippe** [comment]
@qualidafial I am not sure if I follow. You should never depend on 2 different versions of Mockito. If we keep the package names the same, you just have to remove the extra artifact and everything will work.

**Comment by qualidafial** [comment]
We used this strategy in Jdbi from the 2.x to 3.x migration.

- Jdbi 2.x lives at Maven coordinates `org.jdbi:jdbi`, with a root package `org.skife.jdbi.v2`.
- Jdbi 3.x lives at `org.jdbi:jdbi3-<module name>`, with a root package `org.jdbi.v3`.

So both Jdbi 2 and 3 can coexist on the same classpath with no conflicts. They are independent and do not interoperate. This allows users on big projects to migrate from v2 to v3 incrementally instead of forcing them to make the switch all at once.

If Mockito 3 used a different root package name _and_ different Maven coordinates than v2, then users could likewise upgrade their tests incrementally.

Of course, this depends on how difficult the upgrade path will be. If we're just talking about a global search and replace, then it's probably fine to leave it at the same coordinates. If there is a fundamental redesign in the APIs, then it's worthwhile to give users that incremental upgrade path.

**Comment by TimvdLippe** [comment]
Ah okay, thanks for the clarification. Yeah we are not planning on making a
fundamental redesign, so I think we are good in that case :)

Op ma 15 okt. 2018 om 18:55 schreef Matthew Hall <notifications@github.com>:

**Comment by mockitoguy** [comment]
> Write ErrorProne refactoring check that automatically moves all runners to Silent, so that companies can gradually upgrade.

Interesting idea! If you can pull it off, let's do it. I don't think we should block the release on this provided we have the migration guide. Using silent rule/runner everywhere is a search+replace job on code.

**Comment by mockitoguy** [comment]
>Move Junit5 extension from subproject to core project.

Cool idea! AFAIR, the only reason we had a separate module for this was Java8. If JUnit4 and JUnit5 on the same classpath are OK, we should be able to include it in mockito-core.

**Comment by TimvdLippe** [comment]
@mockitoguy Given recent feedback in #769, I think we have to polish strictness more before we can flip the switch in Mockito 3.

**Comment by TimvdLippe** [comment]
My local development of Mockito becomes more and more problematic with Java 6. It came to the point that I simply locally switch to 8 and do all development and rely on CI to catch all accidental breakages.

I would like to propose a very minimal Mockito 3 release:
1. Switch to Java 8 (since a lot of enterprise users are still on 8 rather than 11)
2. Remove deprecated APIs
3. Promote incubated APIs to full public API

In other words, I would propose to NOT include any actual feature breaking changes. If we decouple the language upgrade from breaking changes, we can first cleanup our internal APIs before we make changes to future APIs.

@mockito/core please let me know what you think. If you all approve, I can make this 1 PR and we should be able to do this without any beta phase at all.

**Comment by raphw** [comment]
What about strict mocking in JUnit 4 and JUnit 5? I think those are good ideas to introduce.

**Comment by TimvdLippe** [comment]
Strict mocking has several issues right now that I think we should consider as blockers. Moreover, there is a discrepancy between JUnit 4 and 5. I would rather address these concerns before making them the default.

I would mostly strive for different releases to prevent a Mockito 2 beta-phase again. If we try to tackle it all at once, we make it very difficult for us. Jumping to 8 now pays off some maintenance debt. Then we can tackle strict mocking and ship that in a later release.

**Comment by TimvdLippe** [comment]
We have released the first 3.0.0-beta.1 to jcenter: https://twitter.com/MockitoJava/status/1121716503274033152

**Comment by erikjara** [comment]
Any support for Kotlin on Mockito 3? I found a [brainstorming](https://github.com/mockito/mockito/wiki/Mockito-for-Kotlin) since April 2018 but nothing implemented.

Thanks :)

**Comment by black-snow** [comment]
This issue seems a little outdated as *atm* v3.1.13 is out.

**Comment by szpak** [comment]
For most of the points only the version number is outdated. Mockito 3 has no backward incompatible changes (except for dropping support for Java <8 :) ), so the things are to be considered for Mockito 4.
@TimvdLippe WDYT about updating the issue to reflect Mockito 4 instead of 3?

**Comment by TimvdLippe** [comment]
Yes done. Thanks for the pointer!

**Comment by szpak** [comment]
:+1: 

Btw, I marked to original first point as implemented (instead of removing it) to make your [first comment](https://github.com/mockito/mockito/issues/1517#issuecomment-428914687):
> First point: yes please!

more accurate :).

**Comment by NikolayTach** [comment]
`JUnit (3.0)` test cases within test source folder.

**Comment by TimvdLippe** [comment]
I opened #2418 to remove all deprecated APIs from Mockito, effectively launching Mockito 4.

Looking at the original list that @mockitoguy posted, I don't think we should enable strict stubbing by default and remain optional. I would prefer to have more data on its usefulness, because at Google we had to avoid strict stubbing by default, since it was breaking too many tests. I don't think it is worth breaking so many of our users and instead should keep it opt-in for now.

Lastly, the JUnit5 extension can be moved to core in a minor release, as I think we can make that non-breaking. Therefore, I don't think we need to include that change in Mockito 4.

E.g. my proposal is to merge #2418 and then publish Mockito 4 as-is, since we can tackle the other points separately.

**Comment by grimreaper** [comment]
fwiw at my employer enabling StrictStubbing allowed us to delete much unused code and detected multiple useless tests.
While it may be okay to make it opt-in lets make is possible to globally opt-in (perhaps with a different jar or the way mockito-inline works).

**Comment by TimvdLippe** [comment]
Yes I think that is a good idea. But we can make that happen in a minor release by adding a feature, rather than flipping the default and break all of our users' test. Then we can gather feedback that way.
